[TRTLLM-6106][feat] Add support for KVCache transfer from KVCache reuse path

Tabrizian opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Session-based KV-cache transfers with multi-connection support; Python binding to pin per-request cache blocks (pin_blocks).

- Performance
  - Selective, hash-driven block reuse and lookup to reduce KV-cache transfer and allocation overhead.

- Behavior Changes
  - Executor flag now influences block-reuse lifecycle (requests may be terminated vs retained during context transmission).

- API Changes
  - Environment toggle for selective transfer removed; transceiver/transfer interfaces reorganized.

- Tests
  - Added/updated tests for reuse-tree lookups and transceiver/cache transfer flows.

- Chores
  - Build and header cleanup to align with new transfer architecture.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

[JIRA ticket/NVBugs ID/GitHub issue][fix/feat/doc/infra/...] \<summary of this PR\>

For example, assume I have a PR to support a new feature about cache manager for JIRA ticket TRTLLM-1000, it would be like:

[TRTLLM-1000][feat] Support a new feature about cache manager

Or I have a PR to fix a Llama3 accuracy issue:

[https://nvbugs/1234567][fix] Fix Llama3 accuracy issue
-->

## Description

<!--
Please explain the issue and the solution in short.
-->

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

Tabrizian commented:
/bot run

tensorrt-cicd commented:
[PR_Github #12929](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12929/) [ run ] triggered by Bot

Tabrizian commented:
/bot run

tensorrt-cicd commented:
[PR_Github #13033](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13033/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #13033](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13033/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #9736](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/9736/) completed with status: 'FAILURE'

Tabrizian committed:
Initial iteration for supporting block hash transfer

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Add unittest for findBlocksInReuseTreeByHashes

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

fixes

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Switch from hash id to block key

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Add support for blockKeys

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Fix bugs

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Fix accuracy bug and add tests

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

pcastonguay reviewed (commented):
@Tabrizian Do you think this MR could be broken into multiple MRs? It would make it easier to review. Not sure if it's ready to be reviewed yet...

Tabrizian commented:
/bot run --disable-fail-fast

Tabrizian commented:
@coderabbit review

tensorrt-cicd commented:
[PR_Github #18147](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18147/) [ run ] triggered by Bot

Tabrizian commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #18150](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18150/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #18147](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18147/) [ run ] completed with state `ABORTED`

Tabrizian commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #18157](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18157/) [ run ] triggered by Bot

Tabrizian committed:
Address review comments + testing

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Tabrizian commented on cpp/tensorrt_llm/executor/serialization.cpp:
```diff
@@ -2444,4 +2448,38 @@ ModelType Serialization::deserializeModelType(std::istream& is)
     return su::deserialize<ModelType>(is);
 }
 
+// BlockKey (KV cache)
```
Added

Tabrizian commented on cpp/include/tensorrt_llm/batch_manager/kvCacheManager.h:
```diff
@@ -80,6 +80,32 @@ using MmKey = std::pair<std::array<uint8_t, 32>, SizeType32>;
+//! \param allowPartial whether to append a block smaller than `elementsPerBlock` at the end
+//! \return list of blocks
+template <typename T>
+std::list<std::vector<T>> chopVectorIntoBlocks(
+    std::vector<T> const& vec, SizeType32 usableSize, SizeType32 elementsPerBlock, bool allowPartial)
+{
+    TLLM_CHECK_WITH_INFO(
+        usableSize <= static_cast<SizeType32>(vec.size()), "usableSize=%d > %ld=vec.size()", usableSize, vec.size());
+    std::list<std::vector<T>> blockedVectors;
+    auto const vecEnd = vec.begin() + usableSize;
+    for (auto begin = vec.begin(); begin < vecEnd; begin += elementsPerBlock)
+    {
+        auto blockSize = std::min(elementsPerBlock, static_cast<SizeType32>(std::distance(begin, vecEnd)));
+        auto end = begin + blockSize;
+        if (blockSize == elementsPerBlock || allowPartial)
+        {
+            blockedVectors.emplace_back(begin, end);
+        }
+    }
+    return blockedVectors;
+}
+
```
Added unit test for this function.

Tabrizian commented on cpp/tensorrt_llm/batch_manager/cacheFormatter.h:
```diff
@@ -33,10 +33,231 @@
+        runtime::BufferManager const& bufferManager, int32_t indexFromEnd = 0, BlockKey lastBlockKey = {},
+        LlmRequest const* llmRequest = nullptr, bool recordMeasure = false)
+        : mConnections(std::move(connections))
+        , mDataContext(dataContext)
```
fixed.

Tabrizian commented:
/bot run --disable-fail-fast

Tabrizian commented:
@pcastonguay / @Shixiaowei02 I addressed the review comments. Please review when you get a chance. Thanks.

tensorrt-cicd commented:
[PR_Github #18822](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18822/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #14114](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14114/) completed with status: 'FAILURE'

Tabrizian committed:
Fix TRT path

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Tabrizian commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #18845](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18845/) [ run ] triggered by Bot

pcastonguay commented on tensorrt_llm/_torch/pyexecutor/py_executor.py:
```diff
@@ -1866,24 +1867,32 @@ def _handle_responses(self):
-        self._enqueue_responses(new_responses)
         for request in requests_to_terminate:
             self._terminate_request(request)
+        self._enqueue_responses(new_responses)
```
What's the reason for changing the order here?

pcastonguay commented on tensorrt_llm/_torch/pyexecutor/py_executor.py:
```diff
@@ -1866,24 +1867,32 @@ def _handle_responses(self):
                 if request.is_disagg_context_transmission_state:
-                    self.ctx_in_transmission_requests.append(request)
+                    if self.block_reuse_enabled and not self.kv_cache_manager.is_vswa:
+                        requests_to_terminate.append(request)
```
Do we need to pin blocks for that request?

pcastonguay commented on cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp:
```diff
@@ -1045,10 +1025,39 @@ bool WindowBlockManager::blockInRadixTree(BlockPtr const& block)
+    auto blockedUniqueTokens
+        = chopVectorIntoBlocks<UniqueToken>(blockKey.uniqueTokens, blockKey.uniqueTokens.size(), mTokensPerBlock, true);
+    std::vector<BlockKey> blockKeys;
+    for (auto const& blockedUniqueTokens : blockedUniqueTokens)
```
Can we use different variable names here?

Tabrizian committed:
Review comment

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Tabrizian committed:
Fix tests

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

pcastonguay commented:
@chuangz0 @Shixiaowei02 could you please review the kvCacheTransceiver changes? Thanks.

tensorrt-cicd commented:
[PR_Github #19211](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19211/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #14422](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14422/) completed with status: 'FAILURE'

Tabrizian committed:
Fix pp bugs

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Tabrizian committed:
Fix eagle

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Shixiaowei02 commented on cpp/tensorrt_llm/executor/executorImpl.h:
```diff
@@ -79,6 +79,14 @@ class Executor::Impl
+    struct InTransmissionItem
+    {
+        LlmRequestPtr request;
+        std::optional<SizeType32> lastBlockId; // present when reuse enabled and not variable window
```
The logic based on `lastBlockId` in the code context feels a bit obscure. Would it be possible to add a member function or some comments to help with understanding? Thanks!

Tabrizian committed:
fix pp bugs

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Tabrizian committed:
Fix pp bugs

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Tabrizian committed:
remove newly allocated blocks

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Tabrizian committed:
Fix accuracy bug

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Tabrizian committed:
Fix unit test

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Tabrizian committed:
fix

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #19729](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19729/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #14850](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14850/) completed with status: 'FAILURE'

Tabrizian committed:
Fix compile error

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Tabrizian commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #19814](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19814/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #14908](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14908/) completed with status: 'FAILURE'

Tabrizian commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #19894](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19894/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #19896](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19896/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #19894](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19894/) [ run ] completed with state `ABORTED`
<a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14971/" target="_blank">LLM/main/L0_MergeRequest_PR #14971</a> <a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14971//display/redirect" target="_blank">(Blue Ocean)</a> completed with status: ABORTED

Tabrizian committed:
Fix ci

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Tabrizian commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #19993](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19993/) [ run ] triggered by Bot

Tabrizian committed:
Fix ci

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Tabrizian commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #20035](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20035/) [ run ] triggered by Bot

Tabrizian commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #20035](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20035/) [ run ] completed with state `ABORTED`
<a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15093/" target="_blank">LLM/main/L0_MergeRequest_PR #15093</a> <a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15093//display/redirect" target="_blank">(Blue Ocean)</a> completed with status: ABORTED

Tabrizian commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #20083](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20083/) [ run ] triggered by Bot

Tabrizian committed:
review comment

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>

Tabrizian commented on cpp/include/tensorrt_llm/batch_manager/kvCacheManager.h:
```diff
@@ -767,6 +797,12 @@ class WindowBlockManager
         return 0;
     }
 
+    [[nodiscard]] std::optional<std::shared_ptr<KVCacheBlock>> findBlocksInReuseTreeByBlockKey(
```
Fixed.

tensorrt-cicd commented:
[PR_Github #20115](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20115/) [ run ] triggered by Bot

Tabrizian commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #20147](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20147/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20147](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20147/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15189](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15189/) completed with status: 'FAILURE'

Tabrizian commented:
/bot run --disable-fail-fast
